packages/core/src/modules/routing/RecipientModule.ts

```diff
@@ -181,6 +192,60 @@ export class RecipientModule {
     }
   }
 
+  private async getPickupStrategyForMediator(mediator: MediationRecord) {
+    let mediatorPickupStrategy = mediator.pickupStrategy ?? this.agentConfig.mediatorPickupStrategy
+
+    // If mediator pickup strategy is not configured we try to query if batch pickup
+    // is supported through the discover features protocol
+    if (!mediatorPickupStrategy) {
+      const isBatchPickupSupported = await this.isBatchPickupSupportedByMediator(mediator)
+
+      // Use explicit pickup strategy
+      mediatorPickupStrategy = isBatchPickupSupported
+        ? MediatorPickupStrategy.Explicit
+        : MediatorPickupStrategy.Implicit
+
+      // Store the result so it can be reused next time
+      mediator.pickupStrategy = mediatorPickupStrategy
+      await this.mediationRepository.update(mediator)
+    }
+
+    return mediatorPickupStrategy
+  }
+
+  private async isBatchPickupSupportedByMediator(mediator: MediationRecord) {
+    const { protocolUri } = parseMessageType(BatchPickupMessage.type)
+
+    // Listen for response to our feature query
+    const replaySubject = new ReplaySubject(1)
+    this.eventEmitter
+      .observable<AgentMessageProcessedEvent>(AgentEventTypes.AgentMessageProcessed)
+      .pipe(
```

Comment:
I am not too sure what will be listened to here, but these are a lot of operations. Can't we combine the maps and filter? This might make it a bit less readable but performance should go "up" (probably a very very minor increase though).

Reply:
Yeah sure. This method is really only called once per mediator you connect to as s recipient so was not too worried about adding a few extra method calls. rxjs can be hard to read so slicing it into small steps may help future readers (including me!)

Will update